test/test_um2netcdf.py

```diff
@@ -5,7 +5,217 @@
+            m_mule_vars.return_value = mule_vars
+
+            with mock.patch("iris.load") as m_iris_load:
+                section, item = split_item_code(air_temp_cube.item_code)
+                air_temp_cube.attributes = {um2nc.STASH: DummyStash(section, item)}
+                m_iris_load.return_value = [air_temp_cube]
+
+                with mock.patch("iris.fileformats.netcdf.Saver") as m_saver:  # prevent I/O
+                    m_sman = mock.Mock()
+                    m_saver().__enter__.return_value = m_sman
+
+                    um2nc.process(fake_in_path, fake_out_path, std_args)
+
+                    assert m_sman.update_global_attributes.called is False
+                    assert m_saver.write.called is False  # write I/O prevented
+
+
+def test_process_masking(air_temp_cube, precipitation_flux_cube,
+                         heaviside_uv_cube, heaviside_t_cube,
+                         mule_vars, std_args, fake_in_path, fake_out_path):
+    """Run process() with masking cubes."""
+    with mock.patch("mule.load_umfile"):  # ignore m_load_umfile as process_mule_vars is mocked
+        with mock.patch("umpost.um2netcdf.process_mule_vars") as m_mule_vars:
+            m_mule_vars.return_value = mule_vars
+
+            with mock.patch("iris.load") as m_iris_load:
+                # add cube requiring heaviside_t masking to enable both uv & t code branches
+                geo_potential_cube = DummyCube(30297, "geopotential_height")
+
+                cubes = [air_temp_cube, precipitation_flux_cube, geo_potential_cube,
+                         heaviside_uv_cube, heaviside_t_cube]
+
+                for c in cubes:
+                    attrs = {um2nc.STASH: DummyStash(*split_item_code(c.item_code))}
+                    c.attributes = attrs
+                    c.cell_methods = []
+
+                m_iris_load.return_value = cubes
+
+                with mock.patch("iris.fileformats.netcdf.Saver") as m_saver:  # prevent I/O
+                    m_sman = mock.Mock()
+                    m_saver().__enter__.return_value = m_sman
+
+                    # TODO: fix lat/lon & levels requires c.coord attributes
+                    #       use fixtures to add attrs & remove the patches?
+                    with mock.patch("umpost.um2netcdf.fix_latlon_coord") as m_coord:
+                        with mock.patch("umpost.um2netcdf.fix_level_coord") as m_level:
+                            with mock.patch("umpost.um2netcdf.apply_mask") as m_apply_mask:
+                                with mock.patch("umpost.um2netcdf.cubewrite") as m_cubewrite:
+                                    um2nc.process(fake_in_path, fake_out_path, std_args)
+
+                                    assert m_sman.update_global_attributes.called
+                                    assert m_sman.update_global_attributes.call_count == 2
+                                    assert m_saver.write.called is False  # write I/O prevented
```

Comment:
I'm not too clear on how mocking context managers works, but was wandering if the `.write` method be being called from the `m_saver` mock or the `m_sman` mock?

Reply:
Ooops, that's a bug as the `write()` assert should be called on the mock sman. This is an example of mock fragility & testing implementation details. An improvement is `process()` returning all modified cubes for testing. This may not be a bad idea while the implementation & testing is in flux.

In the meantime, an intermediate fix with the correct mocks is in:  https://github.com/ACCESS-NRI/um2nc-standalone/pull/58/commits/386ff1da9f0a6b012f63ef34f7ceaf437084c8c0

Reply:
The intermediate fix looks good!

Reply:
Oops, testing cubes can be handled by checking the data vars after the `process()` call, thus skipping the need to return anything.

Reply:
I'll resolve this as the code is deleted.